exasol/analytics/audit/audit.py

```diff
@@ -23,3 +37,130 @@ def __init__(
             name=TableNameImpl(table_name, SchemaName(db_schema)),
             columns=(BaseAuditColumns.all + additional_columns),
         )
+        self._column_names = [c.name for c in self.columns]
+
+    def augment(self, queries: Iterator[Query]) -> Iterator[str]:
+        """
+        Process the specified queries and intermerge insert statements
+        into the Audit Log if requested:
+
+        * Queries not requesting any entry into the Audit Log are simply executed.
+
+        * Instances of :class:`AuditQuery` are inserted into the Audit Log,
+          optionally including custom audit_fields and a subquery
+          (SelectQueryWithColumnDefinition)
+
+        * Instances of :class:`ModifyQuery` requesting an entry into the Audit
+          Log are wrapped into one insert statement before and one after. The
+          actual modifying query in between remains unchanged.  The insert
+          statements before and after record the timestamp and, if the
+          ModifyQuery represents an INSERT operation, the number of rows in
+          the modified table.
+        """
+        for query in queries:
+            if not query.audit:
+                yield query.query_string
+            elif isinstance(query, AuditQuery):
+                yield self._insert(query)
+            elif isinstance(query, ModifyQuery):
+                yield from self._wrap(query)
+            else:
+                raise TypeError(
+                    f"Unexpected type {type(query)} of query {query.query_string}"
+                )
+
+    def _insert(self, query: AuditQuery) -> str:
+        insert_statement = (
+            InsertStatement(self._column_names, separator=",\n  ")
+            .add_constants(query.audit_fields)
+            .add_scalar_functions(BaseAuditColumns.values)
+        )
+
+        suffix = ""
+        if query.select_with_columns:
+            alias = TableNameImpl("SUB_QUERY")
+            subquery_columns = {
+                c.name.name: ColumnName(c.name.name, alias)
+                for c in query.select_with_columns.output_columns
+            }
+            insert_statement.add_references(subquery_columns)
+            suffix = (
+                "\nFROM VALUES (1) CROSS JOIN\n"
```

Comment:
You still don't need the cross join, except you want to log a message if the query doesn't return anything, however then a cross join doesn't help you either, because then you would need to add another row with union all

Reply:
I am not sure, whether I understood your proposal completely. 
AFAIK, if there is a subquery, it can result in multiple rows that all need to be added to the Audit Log. 
AFAIK for this I would need something like a join.

Reply:
Cross join is removed now.
Please review again.